osgar/lib/quaternion.py

```diff
@@ -36,9 +36,16 @@ def rotate_vector(vector, quaternion):
     return multiply(part1, con)[:-1]
 
 def euler_zyx(quaternion):
+    # https://en.wikipedia.org/wiki/Conversion_between_quaternions_and_Euler_angles
+    # assumption that quaternion is normalized
     x0, y0, z0, w0 = quaternion
+    sqr_size = x0*x0 + y0*y0 + z0*z0 + w0*w0
+    if abs(sqr_size - 1.0) > 0.00001:
+        k = math.sqrt(sqr_size)
+        x0, y0, z0, w0 = x0/k, y0/k, z0/k, w0/k
     ax =  math.atan2(2*(w0*x0+y0*z0), 1-2*(x0*x0+y0*y0))
-    ay =  math.asin(2*(w0*y0-z0*x0))
+    v = min(1.0, max(-1.0, 2*(w0*y0-z0*x0)))
```

Comment:
If we simply force the value using the min/max code to be within the -1,1 range, aren't we going to skew the reading? Shouldn't we change the other numbers in some proportion to this change also?

Reply:
Also, it is my understanding that a value out of this range on input is actually an invalid input value. So while we should try to work with it instead of crashing with a math exception, we should also indicate in some way (debug log?) that this has occurred.

Reply:
Please see my comment https://github.com/robotika/osgar/pull/532#discussion_r434106515 to similar question